python/taichi/lang/impl.py

```diff
@@ -427,6 +427,9 @@ def make_constant_expr(val):
             return Expr(_ti_core.make_const_expr_f32(val))
         elif pytaichi.default_fp == f64:
             return Expr(_ti_core.make_const_expr_f64(val))
+        elif pytaichi.default_fp == f16:
+            # Use f32 to interact with python
```

Comment:
This will affect the type system because the constant is marked as `f32` in this implementation. Are we only using `float16` as storage types and using at least `float32` for all calculations? If so, this is fine.

Reply:
float16 is a storage type, but also a computation type for some ops e.g. add/mul on cuda. It's indeed a bit tricky since we don't have a native f16 type in python/c++. We can discuss this later, fp16 might need a bit tweak here. :P